consensus/core/src/transaction_certifier.rs

```diff
@@ -0,0 +1,814 @@
+/// necessarily true though, because fastpath execution is only an optimistic latency optimization.
+/// The implementation here takes advantage of this fact. By narrowing the scope of fastpath
+/// execution, post-commit finalization can be made simpler, more efficient and more robust.
+#[derive(Clone)]
+pub(crate) struct TransactionCertifier {
+    // The state of blocks being voted on and certified.
+    certifier_state: Arc<RwLock<CertifierState>>,
+    // The state of the DAG.
+    dag_state: Arc<RwLock<DagState>>,
+    // An unbounded channel to send certified blocks to the block handler
+    // as part of the consensus output.
+    certified_blocks_sender: UnboundedSender<CertifiedBlocksOutput>,
+}
+
+impl TransactionCertifier {
+    pub(crate) fn new(
+        context: Arc<Context>,
+        dag_state: Arc<RwLock<DagState>>,
+        certified_blocks_sender: UnboundedSender<CertifiedBlocksOutput>,
+    ) -> Self {
+        Self {
+            certifier_state: Arc::new(RwLock::new(CertifierState::new(context))),
+            dag_state,
+            certified_blocks_sender,
+        }
+    }
+
+    /// Process own votes on input blocks and votes contained in the input blocks.
+    /// Newly certified blocks are sent to the output channel.
+    ///
+    /// NOTE: blocks from commit sync do not need to be added, because transactions
+    /// arrived via commit sync should not need to be executed on the fastpath.
+    // TODO: add own proposed blocks, even though the contained votes have been processed.
+    pub(crate) fn add_voted_blocks(
+        &self,
+        voted_blocks: Vec<(VerifiedBlock, Vec<TransactionIndex>)>,
+    ) {
+        let certified_blocks = self.certifier_state.write().add_voted_blocks(voted_blocks);
+        if certified_blocks.is_empty() {
+            return;
+        }
+
+        // Only send a block at round R to execute on fastpath if the next round to propose is <= R+2.
+        // So if a quorum of authorities sends a block to fastpath and returns the effect digests to a client,
+        // the quorum is also promising they can only propose at round <= R+2. Then every block at round R+3
+        // must have a link to a certificate of the fastpath block.
+        let next_propose_round = self.dag_state.read().next_propose_round();
+        let certified_blocks: Vec<_> = certified_blocks
+            .into_iter()
+            .filter(|b| b.block.round() + 2 >= next_propose_round)
+            .collect();
+        if certified_blocks.is_empty() {
+            return;
+        }
+
+        if let Err(e) = self.certified_blocks_sender.send(CertifiedBlocksOutput {
+            blocks: certified_blocks,
+        }) {
+            tracing::warn!("Failed to send certified blocks: {:?}", e);
+        }
+    }
+
+    /// Updates the GC round and cleans up obsolete internal state.
+    // TODO: use GC to clean up CertifierState.
+    #[allow(unused)]
+    pub(crate) fn update_gc_round(&self, gc_round: Round) {
```

Comment:
Curious why you have the gc round as an arg, can't you just get it from dag_state?

Reply:
Good point!